File: 0206_reverse_linked_list/solution02.py
from typing import Optional 

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# runtime: 100.00% memory: 52.01% 
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        next = None
        curr = head

        while curr:
            next = curr.next
            curr.next = prev

            prev = curr
            curr = next

        return prev

        

File: 0206_reverse_linked_list/solution02.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

// runtime: 100.00%, memory: 12.86% 
struct ListNode* reverseList(struct ListNode* head) {
    if (head == NULL || head->next == NULL) 
        return head;
    
    struct ListNode *newHead = reverseList(head->next);
    head->next->next = head;
    head->next = NULL;

    return newHead;
}